### PR TITLE
created own TreeMap

### PR DESCRIPTION
**Pull request by dani3lz**

### CODE REVIEW

File: src/main/java/com/endava/internship/collections/Student.java [hunk 1]
```diff
@@ -23,7 +35,25 @@ public Student(String name, LocalDate dateOfBirth, String details) {
 
     public String getDetails() { return details; }
 
-    /*
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
```
Comment (xolari) [inline src/main/java/com/endava/internship/collections/Student.java:40]: Please study style guide and format your code according to it (not only this line and rule).
https://google.github.io/styleguide/javaguide.html

> Braces are used with if, else, for, do and while statements, even when the body is empty or contains only a single statement.

File: src/main/java/com/endava/internship/collections/StudentMap.java [hunk 2]
```diff
@@ -1,78 +1,207 @@
...
+import java.util.*;
 
 public class StudentMap implements Map<Student, Integer> {
+    private int size;
+    private Node parent;
+    private Node found;
+    boolean markForContainsDelete;
+
+
+    private boolean findNodeKey(Node parent, Student student){
```
Comment (xolari) [inline src/main/java/com/endava/internship/collections/StudentMap.java:12]: Ctrl+Alt+L to properly format your code.

File: src/main/java/com/endava/internship/collections/Student.java [hunk 3]
```diff
@@ -1,16 +1,28 @@
...
 import java.time.LocalDate;
+import java.util.Comparator;
+import java.util.Objects;
+import java.util.TreeMap;
 
 /**
  * The class that defines the element that will be contained by your collection
  */
-public class Student //TODO consider implementing any interfaces necessary for your collection
+public class Student implements Comparable<Student> //TODO consider implementing any interfaces necessary for your collection
```
Comment (xolari) [inline src/main/java/com/endava/internship/collections/Student.java:11]: When done, todo comments can be removed.

File: src/main/java/com/endava/internship/collections/StudentMap.java [hunk 4]
```diff
@@ -1,78 +1,207 @@
...
+        return findNodeKey(parent.right, student);
+    }
+
+    private boolean foundNode(Node parent, Student student){
+        if(parent == null) return false;
+        this.found = parent;
+        if(student.compareTo(parent.key) == 0) return true;
+        if(foundNode(parent.left, student)) return true;
+        return foundNode(parent.right, student);
+    }
```
Comment (xolari) [inline src/main/java/com/endava/internship/collections/StudentMap.java:25]: Please move your private methods down to make order logical.

File: src/main/java/com/endava/internship/collections/Student.java (hunk 1 above)
Comment (dani3lz) [reply src/main/java/com/endava/internship/collections/Student.java:33]: Done

File: src/main/java/com/endava/internship/collections/StudentMap.java (hunk 2 above)
Comment (dani3lz) [reply src/main/java/com/endava/internship/collections/StudentMap.java]: Done

File: src/main/java/com/endava/internship/collections/Student.java (hunk 3 above)
Comment (dani3lz) [reply src/main/java/com/endava/internship/collections/Student.java]: Done

File: src/main/java/com/endava/internship/collections/StudentMap.java (hunk 4 above)
Comment (dani3lz) [reply src/main/java/com/endava/internship/collections/StudentMap.java]: Done

File: src/main/java/com/endava/internship/collections/StudentMap.java [hunk 5]
```diff
@@ -1,78 +1,221 @@
...
-        //TODO
-        return false;
+        return size == 0;
     }
 
     @Override
     public boolean containsKey(Object o) {
-        //TODO
-        return false;
+        return findNodeKey(this.parent, (Student) o);
```
Comment (xolari) [inline src/main/java/com/endava/internship/collections/StudentMap.java:22]: It'll throw ClassCastException if not Student provided when desired behavior is return false.

File: src/main/java/com/endava/internship/collections/StudentMap.java [hunk 6]
```diff
@@ -1,78 +1,221 @@
...
 public class StudentMap implements Map<Student, Integer> {
+    private int size;
+    private Node parent;
+    private Node found;
+
     @Override
     public int size() {
-        //TODO
-        return 0;
+        return this.size;
```
Comment (xolari) [inline src/main/java/com/endava/internship/collections/StudentMap.java:12]: No need to use this here and in other methods.

File: src/main/java/com/endava/internship/collections/StudentMap.java [hunk 7]
```diff
@@ -1,78 +1,221 @@
...
-        //TODO
-        return false;
+        return findNodeValue(this.parent, (Integer) o);
     }
 
     @Override
     public Integer get(Object o) {
-        //TODO
-        return null;
+        Student student = (Student) o;
```
Comment (xolari) [inline src/main/java/com/endava/internship/collections/StudentMap.java:32]: ClassCastException can be thrown.

File: src/main/java/com/endava/internship/collections/StudentMap.java [hunk 8]
```diff
@@ -1,78 +1,221 @@
...
+            return false;
+        }
+        this.found = parent;
+        if (student.compareTo(parent.key) == 0) {
+            return true;
+        } else if (foundNode(parent.left, student)) {
+            return true;
+        } else {
+            return foundNode(parent.right, student);
+        }
```
Comment (xolari) [inline src/main/java/com/endava/internship/collections/StudentMap.java:146]: These two methods are mostly identical.

File: src/main/java/com/endava/internship/collections/StudentMap.java [hunk 9]
```diff
@@ -1,78 +1,221 @@
...
+        } else if (result > 0) {
+            parent.right = insert(parent.right, student, value);
+        }
+        return parent;
+    }
+
+    private Student minValue(Node root) {
+        Student minStudent = root.key;
+        while (root.left != null) {
+            minStudent = root.left.key;
```
Comment (xolari) [inline src/main/java/com/endava/internship/collections/StudentMap.java:177]: No need to assign value each time to minStudent if in the end you can get this data from minStudent.

File: src/main/java/com/endava/internship/collections/StudentMap.java (hunk 6 above)
Comment (dani3lz) [reply src/main/java/com/endava/internship/collections/StudentMap.java]: Done

File: src/main/java/com/endava/internship/collections/StudentMap.java (hunk 8 above)
Comment (dani3lz) [reply src/main/java/com/endava/internship/collections/StudentMap.java]: Done

File: src/main/java/com/endava/internship/collections/StudentMap.java (hunk 9 above)
Comment (dani3lz) [reply src/main/java/com/endava/internship/collections/StudentMap.java]: Done

File: src/main/java/com/endava/internship/collections/StudentMap.java [hunk 10]
```diff
@@ -1,78 +1,201 @@
...
+            return null;
+        }
+        if (student.compareTo(parent.key) == 0) {
+            found = parent;
+            return found;
+        } else if (foundNode(parent.left, student) != null) {
+            return found;
+        } else if (foundNode(parent.right, student) != null) {
+            return found;
+        }
```
Comment (xolari) [inline src/main/java/com/endava/internship/collections/StudentMap.java:131]: Combine these two conditions, they have the same body.
If you return in the previous return block, it's not necessary to write else before next if.

Comment (dani3lz) [reply src/main/java/com/endava/internship/collections/StudentMap.java]: Done

**Review by xolari — APPROVED** [review]

**Pull request merged**